### ISSUE TITLE
Implementation error

### ISSUE THREAD

**Issue by JacekWydra** [davda54/sam#16]
Hi,
I am pretty sure that there is an error in the implementation of _grad_norm method. 
If we look at the equation 2 in the original paper the denominator contains q-norm of vector grad(**w**) to the power of q and all that to the power 1/p. In the case of q=2 and p=2  we get (L2(grad(**w**)) ^ 2) ^ 0.5. 
Your implementation stacks L2 norms of each parameter tensor, stacks them and calculate L2 norm of those stacked norm which is not the same as the above formula. You should ravel and concatenate all the parameter tensors and then calculate L2 norm two times.

Please check my reasoning but I am almost sure, that I am right.
Best Regards,
Jacek

**Comment by davda54** [comment]
Hi,

you are right taking the norm of concatenated parameters would work but it is a very costly operation for big models. This is why I divide the computation and calculate the L2 norm separately for each layer. I believe that this optimization is correct, do you have any counterexample?

I don't want to make a rigorous proof but here is an example to show you the intuition behind that code: notice that
<img src="https://latex.codecogs.com/png.latex?\sqrt{a^2%20+%20b^2%20+%20c^2%20+%20d^2}%20=%20\sqrt{(\sqrt{a^2%20+%20b^2})^2%20+%20(\sqrt{c^2%20+%20d^2})^2}">

On top of that, even PyTorch uses the same trick for [gradient clipping](https://github.com/pytorch/pytorch/blob/e6779d4357ae94cc9f9fedb83a87eb6126016769/torch/nn/utils/clip_grad.py#L36) :)

Cheers,
David

**Comment by JacekWydra** [comment]
Hi David,
my bad, thanks for clarification. I did not notice that those are equal. 
Have a nice day :)
Cheers,
Jacek

**Comment by shuo-ouyang** [comment]
Hi, I am confused that should we calculate grad norm for each parameter separately or calculate it for all parameters?

**Comment by GeoffreyChen777** [comment]
> Hi, I am confused that should we calculate grad norm for each parameter separately or calculate it for all parameters?

Same question. Do you have any insight?

**Comment by davda54** [comment]
We calculate the norm of the gradients from all parameters, as explained above.
